core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -131,68 +128,72 @@
    * @param otherFields other fields to index the set
    */
   public IndexedSet(FieldIndex<T> field, FieldIndex<T>... otherFields) {
-    mIndexMap = new HashMap<FieldIndex<T>, Integer>(otherFields.length + 1);
-    mIndexMap.put(field, 0);
+    Map<FieldIndex<T>, ConcurrentHashMap<Object, ConcurrentHashSet<T>>> indexMap =
+        new HashMap<>(otherFields.length + 1);
+    indexMap.put(field, new ConcurrentHashMap<Object, ConcurrentHashSet<T>>());
     for (int i = 1; i <= otherFields.length; i++) {
-      mIndexMap.put(otherFields[i - 1], i);
-    }
-
-    mSetIndexedByFieldValue = new ArrayList<Map<Object, Set<T>>>(mIndexMap.size());
-    for (int i = 0; i < mIndexMap.size(); i++) {
-      mSetIndexedByFieldValue.add(new HashMap<Object, Set<T>>());
+      indexMap.put(otherFields[i - 1], new ConcurrentHashMap<Object, ConcurrentHashSet<T>>());
     }
+    // read only, so it is thread safe and allows concurrent access.
+    mIndexMap = Collections.unmodifiableMap(indexMap);
   }
 
   /**
    * Removes all the entries in this set.
    */
+  // TODO(gpang): clear() does not play well with other concurrent operations. Currently, it is
+  // ok, since the only use of clear() is a context without concurrent access.
   public void clear() {
-    synchronized (mLock) {
-      mObjects.clear();
-      for (Map<Object, Set<T>> mapping : mSetIndexedByFieldValue) {
-        mapping.clear();
-      }
+    // Clear the indexes first, so objects cannot be found via index.
+    for (Map<Object, ConcurrentHashSet<T>> mapping : mIndexMap.values()) {
+      mapping.clear();
     }
+    // Remove all the objects.
+    mObjects.clear();
```

Comment:
A concrete concurrency bug I came up:

Thread-1 add(newObj), executing to just add the newObj to mObjects
Thread-2 clear(), executing to `mObjects.clear()` and switched to Thread-1
Thread-1 add newObj to indexes
Thread-2 clear mObjects

At this point, newObj is in indexes but not in mObjects...